aten/src/ATen/native/MathBitsFallback.h

```diff
@@ -72,55 +87,79 @@ struct MathOpFallback {
       return;
     }
 
-    // Mutable inputs to be tracked separately
-    std::vector<Tensor> mutable_inputs;
+    /*
+    Different possible cases:
+    1. Functions with no tensorlist inputs
+      a. no mutable args
+      b. one or more mutable args
+         b.1. shared memory between mutable and non-mutable args (--handled)
+         b.2. shared memory between two or more mutable args (incorrect result but this is bad
+              and users shouldn't do it anyway)
+         b.3. shared memory between two or more non-mutable args (works fine since we never modify the memory)
+         b.4. no shared memory between args (-- works fine)
+    2. Functions with tensorlist inputs
+      a. no mutable args (-- works fine)
+      b. Mutable tensor arg(s) but non-mutable tensorlist arg
+        b.1. All the possible cases listed in 1.b (-- works fine)
+        b.2. shared memory between a tensor arg and a tensorlist arg ( -- Not currently supported)
+      c. Mutable tensorlist arg(s) ( -- Not currently supported)
+        c.1. shared memory between a mutable and non-mutable tensorlist
+        c.2. shared memory between two or more mutable tensorlist args
+        c.3. shared memory between two or more non-mutable tensorlist args
+        c.4. shared memory between two or more non-mutable tensor args
+        c.5. shared memory between a tensor and tensorlist arg
+        ...
+        c.n. no shared memory between args
+    */
 
+    // updates for non-mutable inputs
     for (const auto i : c10::irange(num_arguments)) {
       auto& ivalue = (*stack)[stack_start + i];
       if (!(ivalue.isTensor() || ivalue.isTensorList())) {
         continue;
       }
       const auto& argument = arguments[i];
-      bool mut_arg = false;
       if (argument.alias_info()) {
         // Was already tested by is_write loop above
         TORCH_INTERNAL_ASSERT_DEBUG_ONLY(argument.alias_info()->isWrite());
-        mut_arg = true;
+        continue;
       }
       if (ivalue.isTensor()) {
-        if (!is_bit_set(ivalue.toTensor())) {
-          continue;
-        }
-
         auto tensor = std::move(ivalue).toTensor();
         TORCH_CHECK_NOT_IMPLEMENTED(!tensor.is_meta(), op_name, " fallback does not support meta tensors.");
-        if (mut_arg) {
-          // TODO: This is a waste if the argument is write only
-          _set_bit(tensor, false);
-          math_op_(tensor);
-          mutable_inputs.emplace_back(tensor);
+        if (check_for_alias_with_mut_arg) {
+          bool cloned = false;
+          for (const auto& mutable_input : mutable_inputs) {
+            if (tensor.is_alias_of(mutable_input)) {
+              tensor = tensor.clone();
```

Comment:
I have an alternate proposal. Right now, you have:

```
// precondition: self is conjugate, self aliases with other
// what currently happens
other = other.clone()
self.set_bit_(false)
math_op_(self)

redispatch_(self, other)

self.set_bit_(true)
math_op_(self)
```

Here is the proposed alternative:

```
// precondition: self is conjugate (no condition on other)
// other possibly policy
self_clone = self.clone()
redispatch_(self_clone, resolve_conj(other))
self.copy_(self_clone)
```

This is quite a lot simpler, and even better, it doesn't require the aliasing test. It is slightly worse, performance wise, in the case where self is conjugate but other doesn't alias self; in this case, we will do an unnecessary copy of self to clone to do the conjugate operation on it (indeed, this is precisely what the old optimization was about optimizing). But it doesn't seem worth the ensuing complication to detect aliases.

Reply:
Here's roughly the perf comparison for the two scenarios:
CPU:
```
M=torch.randn(8, dtype=torch.cdouble).conj() # mutable conjugate view tensor
N=torch.randn(8, dtype=torch.cdouble) # f(self_clone, args)

[-------------- Effective materialization of conj --------------]
                          |  Using clone  |  Using conj_physical_
1 threads: ------------------------------------------------------
      M.clone()           |     2385.4    |
      M.conj_physical_()  |               |         737.8

Times are in nanoseconds (ns).

[-------------- Effective materialization of conj --------------]
                          |  Using copy_  |  Using conj_physical_
1 threads: ------------------------------------------------------
      M.copy_(N)          |     921.1     |
      M.conj_physical_()  |               |         620.6

Times are in nanoseconds (ns).
```
CUDA:
```
[-------------- Effective materialization of conj --------------]
                          |  Using clone  |  Using conj_physical_
1 threads: ------------------------------------------------------
      M.clone()           |      14.3     |
      M.conj_physical_()  |               |          7.6

Times are in microseconds (us).

[-------------- Effective materialization of conj --------------]
                          |  Using copy_  |  Using conj_physical_
1 threads: ------------------------------------------------------
      M.copy_(N)          |      5.9      |
      M.conj_physical_()  |               |          4.4

Times are in microseconds (us).
```
Aliasing checks are certainly far cheaper than always conjugating since we only check if the two tensors in question share storage or not (2 * num_args worst case since we usually don't have more than two mutable argument tensors, and in the worst case all argument tensors are conjugated). 

It's also true that the second approach will simplify the logic (although we'll have to track the mut_arg_clone tensors separately in this case). The perf offset for cloning and copying instead of conjugating in-place twice is 1.7 us and 8-9 us on CPU and CUDA respectively [fyi: the perf difference on CPU is almost invariant of the size]. 

I think the difference on CUDA seems pretty significant to forego the first approach in favor of the latter.

Reply:
This is a significant perf difference (50% or more) for some ops like `torch.{dot/vdot}` see here: https://github.com/pytorch/pytorch/pull/62915#issuecomment-895338209

Reply:
> This is a significant perf difference (50% or more) for some ops like torch.{dot/vdot} see here: #62915 (comment)

This isn't relevant, because you're going to  bypass MathBitFallback in that case right?

Reply:
I agree that the simplified variant is slower. But everything is slow *anyway* because you're hitting the fallback case (and materializing a bunch of stuff). So the question is how much the perf improvement (in an already slow case) balances with complexity. Well, I suppose that if you really want this optimization, we can keep it, but then I need to rereview the algorithm again.

Reply:
ok let's get the current version in for 1.10 and we can revisit this after